### PR TITLE
Refactor infrastructure to facilitate commissioning and SV

### PR DESCRIPTION
**Pull request by geordie666**
Refactor to allow separate and additional commissioning and SV target selections. Includes:

- Added `sv` and `commissioning` directories.
- New infrastructure to have different cuts for SV and commissioning:
    - separate target masks (e.g. `sv/data/sv_targetmask.yaml`).
    - separate cuts modules (e.g. `sv_cuts.py`; not yet populated with actual algorithms).
- Added executables for SV/commissioning (e.g. `select_sv_targets`).
- The initial `NUMOBS` and `PRIORITY` are now included as columns in the `targets-` files.
- The initial `NUMOBS` is now hardcoded in the target masks, instead of being set by MTL.
- `SV` bits added to the target masks to determine if targets are from SV/comm/main survey.
- sv/comm/main can now be written to the header of the `targets-` files.
- `SUBPRIORITY` is set when writing, not reading, targets to facilitate reproducibility.

This should facilitate different target selections and MTL for SV and commissioning. As a bonus it also fixes https://github.com/desihub/desitarget/issues/337.

### CODE REVIEW

**Comment by geordie666** [comment]
@moustakas: This does everything we discussed this week. It doesn't yet have the actual selection algorithms in the`cuts.py` files, but all of the infrastructure should be in place. 

An example output file with `PRIORITY` and `NUMOBS` included as columns is at `/global/cscratch1/sd/adamyers/targets-PR346.fits`

File: py/desitarget/targets.py [hunk 1]
```diff
@@ -289,6 +289,92 @@ def decode_targetid(targetid):
 
     return objid, brickid, release, mock, sky
 
+############################################################
+def initial_priority_numobs(targets):
+    """highest initial priority and corresponding numobs for an array of target bits
```
Comment (moustakas) [inline py/desitarget/targets.py]: Just reading the code, this looks like it's doing the right thing, but I'll need to test-drive it and will likely have more input after the merge.

**Comment by geordie666** [comment]
I'll get to your review later. But...note that I haven't done anything in `commissioning_cuts.py`, so you should just ignore the substantial amount of code in there, for now. It's a placeholder.

File: py/desitarget/commissioning/data/commissioning_targetmask.yaml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+#- Default priority for "MORE_ZGOOD" is same as "UNOBS"
+#- Other states are default 1 unless specified here
+#- -1 means that the concept of priority doesn't apply to this bit
+priorities:
+    commissioning_mask:
+        STD_GAIA: -1
+        STD_BRIGHT: -1
+        STD_TEST: -1
+        STD_CALSPEC: -1
+        SV_BGS: {UNOBS: 2100, MORE_ZWARN: 2100, MORE_ZGOOD: 1000, DONE: 2, OBS: 1, DONOTOBSERVE: 0}
```
[the remark this answers is not in the document]
Comment (sbailey) [reply py/desitarget/commissioning/data/commissioning_targetmask.yaml]: Note: Standard star priorities will require coordination with fiberassignment, which currently treats standards as a boolean: either they are or they aren't.  We considered using SUBPRIORITY to rank them and then talked ourselves out of that.  For WD we seem to be leaning toward just making them high priority normal targets to guarantee placement.  And now CALSPEC standards come up as another case where we might want to prioritize...

Bottom line: some form of supporting standard star priorities seems necessary, but let's revisit that in a different PR and not get bogged down here.  I think this infrastructure allows desitarget to assign standard star priorities; it is more a matter of how that gets used downstream in fiberassign.

[the remark this answers is not in the document]
Comment (moustakas) [reply py/desitarget/commissioning/data/commissioning_targetmask.yaml]: Sorry for not being clear:  I was saying that for *commissioning* only these targets should be "science", not calibration targets, which would enable us to prioritize, e.g., CALSPEC over other standards without needing any more code.  In SV and the "main" survey I don't think we'll need to prioritize standards (other than WDs, which we've talked about -- but have implemented -- that they'll become high-priority science targets themselves).

File: bin/select_commissioning_targets [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    log.critical('no sweep or tractor files found')
+    sys.exit(1)
+
+log.info("running on {} processors".format(ns.numproc))
+
+#ADM if specific bit names were passed, use them, otherwise run all target classes
+tcnames = ["ELG", "QSO", "LRG", "MWS", "BGS", "STD"]
+if ns.tcnames is not None:
+    tcnames = [ bn for bn in ns.tcnames.split(',') ]
+
```
[the remark this answers is not in the document]
Comment (moustakas) [reply bin/select_commissioning_targets]: And for the record: in `run_target_qa` you use the argument `bitnames` for the list of desired targets (QSO, ELG, etc.).  Would be good to use a common input for both target selection and QA.

**Comment by moustakas** [comment]
> * The initial NUMOBS and PRIORITY are now included as columns in the targets- files.
> * The initial NUMOBS is now hardcoded in the target masks, instead of being set by MTL.

@geordie666 I don't know how far along you are on this PR, but I wonder if it would be possible to merge in the preceding changes (perhaps as a fast-track separate PR)?  I'm updating `desitarget/bin/select_mock_targets` and these changes are mission-critical (whereas I can do the SV/commissioning updates later, separately).

Thanks.

File: bin/select_commissioning_targets (hunk 3 above)
[the remark this answers is not in the document]
Comment (geordie666) [reply bin/select_commissioning_targets]: I caught things not in the list. I don't want to change `tcnames` to `targets` as targets is frequently used as a variable throughout the code. I will change `bitnames` in `QA` to tcnames, in a later PR though.

File: bin/select_commissioning_targets [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+from argparse import ArgumentParser
+ap = ArgumentParser(description='Generates DESI commissioning target bits from Legacy Surveys sweeps or tractor files')
+ap.add_argument("src", 
+                help="Tractor/sweeps file or root directory with tractor/sweeps files")
+ap.add_argument("dest", 
+                help="Output target selection file")
+ap.add_argument("--gaiadir", 
+                help="Base directory for chunked Gaia files (defaults to ['project/projectdirs/cosmo/work/gaia/chunks-gaia-dr2-astrom'] for NERSC)",
+                default='/project/projectdirs/cosmo/work/gaia/chunks-gaia-dr2-astrom')
+ap.add_argument("--gaiamatch", action='store_true',
```
[the remark this answers is not in the document]
Comment (geordie666) [reply bin/select_commissioning_targets]: I'm confused as to what you mean by "hard-coded" here. I have carefully attempted to not hard-code directories. All of them can be passed as input flags or variables in functions. This allows a softer use where you can have subsets of files in different directories while testing the code.

I think what you mean is that you'd actually *prefer* these directories *to be hard-coded* using environment variables. I agree with that in some places, but not until we establish a top-level environment variable for the data. And, not for data products that are being borrowed from the imaging surveys. I'm hoping that the Gaia-matching will be deprecated in DR8, so I don't really want to set up an environment variable and "official" desitarget location for the Gaia files, when we'll just stop using it. Although, it's a reasonable argument. Maybe we should.

File: bin/select_commissioning_targets [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+#ADM if specific bit names were passed, use them, otherwise run all target classes
+tcnames = ["ELG", "QSO", "LRG", "MWS", "BGS", "STD"]
+if ns.tcnames is not None:
+    tcnames = [ bn for bn in ns.tcnames.split(',') ]
+
+targets = select_targets(infiles, numproc=ns.numproc,
+                         gaiamatch=ns.gaiamatch, gaiadir=ns.gaiadir, tcnames=tcnames)
+
+io.write_targets(ns.dest, targets, indir=ns.src, survey="comm", nside=nside)
+
```
[the remark this answers is not in the document]
Comment (geordie666) [reply bin/select_commissioning_targets]: We settled on `cmx`

File: py/desitarget/commissioning/commissioning_cuts.py [hunk 6]
```diff
@@ -0,0 +1,2524 @@
+"""
+desitarget.comm.comm_cuts
+=========================
```
[the remark this answers is not in the document]
Comment (geordie666) [reply py/desitarget/commissioning/commissioning_cuts.py]: We settled on `cmx`

File: py/desitarget/commissioning/commissioning_cuts.py [hunk 7]
```diff
@@ -0,0 +1,2524 @@
...
+    #ADM and only use the z-band color shift when r and z are non-zero
+    w = np.where( (rflux != 0) & (zflux != 0) )
+    rshift = rflux * 10**(+0.4*0.012)
+    zshift = zflux * 10**(-0.4*0.000)
+
+    rshift[w] = (rflux[w] * 10**(+0.4*0.012) * (rflux[w]/zflux[w])**complex(-0.029)).real
+    zshift[w] = (zflux[w] * 10**(-0.4*0.000) * (rflux[w]/zflux[w])**complex(+0.009)).real
+
+    return gshift, rshift, zshift
+
```
[the remark this answers is not in the document]
Comment (geordie666) [reply py/desitarget/commissioning/commissioning_cuts.py]: When we actually have the cuts, we should import utility functions from `desitarget.cuts` (such as the "shift" function you note) but we should *not* import any of the actual functions that perform the cuts, in case we later change code in `desitarget.cuts` for the main survey.

File: py/desitarget/commissioning/commissioning_cuts.py [hunk 8]
```diff
@@ -0,0 +1,2524 @@
...
+    elg &= zflux < rflux * 10**(1.6/2.5)                       # (r-z)<1.6
+
+    # Clip to avoid warnings from negative numbers raised to fractional powers.
+    rflux = rflux.clip(0)
+    zflux = zflux.clip(0)
+    elg &= rflux**2.15 < gflux * zflux**1.15 * 10**(-0.15/2.5) # (g-r)<1.15(r-z)-0.15
+    elg &= zflux**1.2 < gflux * rflux**0.2 * 10**(1.6/2.5)     # (g-r)<1.6-1.2(r-z)
+
+    return elg
+
```
[the remark this answers is not in the document]
Comment (geordie666) [reply py/desitarget/commissioning/commissioning_cuts.py]: Again, we should import utility functions but we should absolutely *not* import code that is performing the cuts. I'll be working with that philosophy moving forward.

File: py/desitarget/commissioning/commissioning_cuts.py [hunk 9]
```diff
@@ -0,0 +1,2524 @@
...
+        objtype: array_like or None
+            If given, the TYPE column of the Tractor catalogue.
+        release: array_like[ntargets]
+            The Legacy Survey imaging RELEASE (e.g. http://legacysurvey.org/release/).
+        deltaChi2: array_like or None
+             If given, difference in chi2 between PSF and SIMP morphology.
+        primary: array_like or None
+            If given, the BRICK_PRIMARY column of the catalogue.
+
+    Returns
```
[the remark this answers is not in the document]
Comment (geordie666) [reply py/desitarget/commissioning/commissioning_cuts.py]: Again, moving forward, we should attempt to be fragile to changes in code that performs SV/main survey/cmx cuts (which should be different for sv and cmx and should not rely on knowledge of the main survey cuts), but not fragile to changes in utility functions.

File: py/desitarget/commissioning/commissioning_targetmask.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+                            priorities[bitname][state] = 1
+                else:
+                    priorities[bitname] = dict()
+
+        #- add to the extra info dictionary for this target mask
+        for bitdef in _bitdefs[maskname]:
+            bitname = bitdef[0]
+            bitdef[3][handle] = priorities[bitname]
+    return
+
```
[the remark this answers is not in the document]
Comment (geordie666) [reply py/desitarget/commissioning/commissioning_targetmask.py]: I updated the bit-setting code and it is now much more streamlined.
